### PR TITLE
Add Zero1+GA to improve performance

### PR DESCRIPTION
**Pull request by hx89**
Add Zero1+gradient-accumulation (GA) support in Maxtext to reduce the communication overhead and improve performance.

When Zero1+GA is enabled by setting shard_optimizer_over_data=True with DP, we will only have AG before the GA loop and AR after the GA loop, and there are no communication of gemm weights and activations within GA loop.

prevent_cse need to be set to False when GA > 1 to avoid extra ops generated by XLA.

### CODE REVIEW

File: src/MaxText/train_utils.py [hunk 1]
```diff
@@ -274,3 +281,13 @@ def validate_train_config(config):
         " different attention mechanism. With synthetic data, the format is not"
         " important as packing is not applied."
     )
+
+  # Automatically disable shardy when gradient accumulation is enabled
+  if config.gradient_accumulation_steps > 1 and config.shardy:
```
Comment (gobbleturk) [inline src/MaxText/train_utils.py]: can you move this logic to pyconfig?

Comment (gobbleturk) [reply src/MaxText/train_utils.py]: Can you also add hardware_type = "gpu" or "gpu_multiprocess" to this check? This particular issue is only seen on gpu

File: src/MaxText/train.py [hunk 2]
```diff
@@ -210,12 +210,24 @@ def train_step(model, config, state_mesh_shardings, state, data, dropout_rng):
     extra_dpo_args = [reference_params]
     _loss_fn = dpo_loss_fn
 
+  params = state.params
+
   if config.gradient_accumulation_steps > 1:
+    def convert_to_bf16(param):
```
Comment (gobbleturk) [inline src/MaxText/train.py]: Why not use param.astype(jnp.bfloat16) directly?

File: src/MaxText/maxtext_utils.py [hunk 3]
```diff
@@ -938,6 +938,62 @@ def setup_training_state(model, data_iterator, tx, config, rng, mesh, checkpoint
...
+
+    if size % mesh.shape['data'] == 0 and (partition is None or 'tensor' not in partition):
+      added_component = ('data',) + partition
+      new_pspec = jax.sharding.PartitionSpec(*(pspec[:idx] + (added_component,) + pspec[idx+1:]))
+      new_sharding = jax.sharding.NamedSharding(sharding.mesh, new_pspec)
+      # return sharding.with_spec(new_pspec)
+      return new_sharding
+  return sharding
+
+def maybe_update_params_sharding_with_opt(config, state_mesh_shardings):
```
Comment (gobbleturk) [inline src/MaxText/maxtext_utils.py:1023]: can you include a comment what this function does and what it returns?

File: src/MaxText/layers/decoders.py [hunk 4]
```diff
@@ -427,7 +427,7 @@ def map_fn(path, value):
       # Apply remat policy to layer
       layer = nn.remat(
           block_layer,
-          prevent_cse=not self.config.scan_layers,
+          prevent_cse=not (self.config.scan_layers or self.config.gradient_accumulation_steps > 1),
```
Comment (gobbleturk) [inline src/MaxText/layers/decoders.py]: we need an additional check here that "config.hardware==gpu (or gpu_multiprocess)"

Given this is significant logic that is likely to change we should move it into a small function in maxtext utils, e.g.
def prevent_cse(config):
   return not self.config.scan_layers or (self.config.gradient_accumulation_steps > 1 and config.hardware==...)

File: src/MaxText/train.py (hunk 2 above)
Comment (gobbleturk) [reply src/MaxText/train.py]: I guess to handle fp8 guess?

File: src/MaxText/train.py [hunk 5]
```diff
@@ -210,12 +210,24 @@ def train_step(model, config, state_mesh_shardings, state, data, dropout_rng):
     extra_dpo_args = [reference_params]
     _loss_fn = dpo_loss_fn
 
+  params = state.params
+
   if config.gradient_accumulation_steps > 1:
```
Comment (gobbleturk) [inline src/MaxText/train.py:217]: doesn't have to be part of this PR but this makes train very hard to read - I think this amount of logic and code warrants an entire new file gradient_accumulation.py

File: src/MaxText/train.py [hunk 6]
```diff
@@ -210,12 +210,24 @@ def train_step(model, config, state_mesh_shardings, state, data, dropout_rng):
...
+    def convert_to_bf16(param):
+      if param.dtype == jnp.float32:
+        return param.astype(jnp.bfloat16)
+      else:
+        return param
+    # Cast params to bf16 and unshard before gradient accumulation loop so that 
+    # all-gather is done once in bf16 
+    params_bf16 = jax.tree_util.tree_map(convert_to_bf16, params)
```
Comment (gobbleturk) [inline src/MaxText/train.py]: This will only get run with gradient accumulation? Shouldn't this instead only get run when shard_optimizer_over_data is true?

File: src/MaxText/train.py (hunk 2 above)
Comment (hx89) [reply src/MaxText/train.py]: yes it's safer to makes sure the conversion is only done on fp32 param.

File: src/MaxText/maxtext_utils.py [hunk 7]
```diff
@@ -938,6 +938,62 @@ def setup_training_state(model, data_iterator, tx, config, rng, mesh, checkpoint
...
+      partition = ()
+
+    if isinstance(partition, str):
+      partition = (partition,)
+
+    if size % mesh.shape['data'] == 0 and (partition is None or 'tensor' not in partition):
+      added_component = ('data',) + partition
+      new_pspec = jax.sharding.PartitionSpec(*(pspec[:idx] + (added_component,) + pspec[idx+1:]))
+      new_sharding = jax.sharding.NamedSharding(sharding.mesh, new_pspec)
+      # return sharding.with_spec(new_pspec)
```
Comment (NuojCheng) [inline src/MaxText/maxtext_utils.py]: could you remove this line?

File: src/MaxText/train.py (hunk 5 above)
Comment (NuojCheng) [reply src/MaxText/train.py:217]: Agree. I will move them in a separately PR.

**Comment by hx89** [comment]
Thanks for reviewing! I updated the PR to address the comments.
